hazelcast/src/main/java/com/hazelcast/spi/impl/classicscheduler/ClassicOperationScheduler.java

```diff
@@ -0,0 +1,432 @@
+        for (int threadId = 0; threadId < threads.length; threadId++) {
+            String threadName = threadGroup.getThreadPoolNamePrefix("partition-operation") + threadId;
+            LinkedBlockingQueue workQueue = new LinkedBlockingQueue();
+            ConcurrentLinkedQueue priorityWorkQueue = new ConcurrentLinkedQueue();
+            PartitionOperationThread operationThread = new PartitionOperationThread(
+                    threadName, threadId, workQueue, priorityWorkQueue, logger,
+                    threadGroup, nodeExtension, partitionOperationHandlers);
+            threads[threadId] = operationThread;
+            operationThread.start();
+        }
+
+        return threads;
+    }
+
+    private GenericOperationThread[] initGenericThreads() {
+        // we created as many generic operation handlers, as there are generic threads.
+        int threadCount = genericOperationHandlers.length;
+        GenericOperationThread[] threads = new GenericOperationThread[threadCount];
+        for (int threadId = 0; threadId < threads.length; threadId++) {
+            String threadName = threadGroup.getThreadPoolNamePrefix("generic-operation") + threadId;
+            OperationHandler operationHandler = genericOperationHandlers[threadId];
+            GenericOperationThread operationThread = new GenericOperationThread(
+                    threadName, threadId, genericWorkQueue, genericPriorityWorkQueue,
+                    logger, threadGroup, nodeExtension, operationHandler);
+            threads[threadId] = operationThread;
+            operationThread.start();
+        }
+
+        return threads;
+    }
+
+    @Override
+    public OperationHandler[] getPartitionOperationHandlers() {
+        return partitionOperationHandlers;
+    }
+
+    @Override
+    public OperationHandler[] getGenericOperationHandlers() {
+        return genericOperationHandlers;
+    }
+
+    @Override
+    public boolean isAllowedToRunInCurrentThread(Operation op) {
+        return isAllowedToRunInCurrentThread(op.getPartitionId());
+    }
+
+    @Override
+    public boolean isOperationThread() {
+        return Thread.currentThread() instanceof OperationThread;
+    }
+
+    @Override
+    public boolean isInvocationAllowedFromCurrentThread(Operation op) {
+        return isInvocationAllowedFromCurrentThread(op.getPartitionId());
+    }
+
+    private boolean isAllowedToRunInCurrentThread(int partitionId) {
+        Thread currentThread = Thread.currentThread();
+
+        // IO threads are not allowed to run any operation
+        if (currentThread instanceof NIOThread) {
+            return false;
+        }
+
+        //todo: do we want to allow non partition specific tasks to be run on a partitionSpecific operation thread?
+        if (partitionId < 0) {
+            return true;
+        }
+
+        //we are only allowed to execute partition aware actions on an OperationThread.
```

Comment:
Doesn't this comment fit better to the previous `if (partitionId < 0)`? The actual if about the thread type, not if the operation is partition aware (and yeah, this is an old issue if it is one).